File: c/stack/maximal-rectangle.c
int *st;

int maxArea(int *heights, int matrixColSize)
{
    int top = -1;
    int i;
    int area = 0;

    for(i = 0; i <= matrixColSize; ++i){
        while(top != -1 && heights[i] <= heights[st[top]]){
            int h = heights[st[top--]];
            int w = ((top == -1) ? i : (i - st[top] - 1));

            if(h * w > area)
                area = h * w;
        }
        st[++top] = i;
    }

    return area;
}

int maximalRectangle(char** matrix, int matrixRowSize, int matrixColSize) {
    int *heights;
    int i,j;
    int max_area = 0;
    int area = 0;

    heights = malloc(sizeof(int) * (matrixColSize+1));
    memset(heights, 0, sizeof(int) * (matrixColSize+1));
    st = malloc(sizeof(int) * (matrixColSize+1));

    for(i = 0; i < matrixRowSize; ++i){
        for(j = 0; j < matrixColSize; ++j){
            if(matrix[i][j] == '0')
                heights[j] = 0;
            else
                heights[j]++;
        }

        area = maxArea(heights,matrixColSize);
        max_area = area > max_area ? area : max_area;
    }

    free(heights);
    free(st);

    return max_area;
}

File: python/stack/maximal-rectangle.py
class Solution(object):
    def maxArea(self, heights):
        area = 0
        st = []

        for i in range(len(heights)):
            while st and heights[i] <= heights[st[-1]]:
                h = heights[st[-1]]
                st.pop()
                w = i if not st else (i - st[-1] - 1)

                if h * w > area:
                    area = h * w

            st.append(i)

        return area

    def maximalRectangle(self, matrix):
        if not matrix:
            return 0

        heights = [0] * (len(matrix[0])+1)
        max_area = 0;

        for row in matrix:
            for j in range(len(row)):
                if row[j] == '1':
                    heights[j] += 1
                else:
                    heights[j] = 0;

            area = self.maxArea(heights)
            if area > max_area:
                max_area = area


        return max_area
